node/libs/roles/src/validator/messages/committee.rs

```diff
@@ -0,0 +1,212 @@
+    pub fn get(&self, index: usize) -> Option<&WeightedValidator> {
+        self.vec.get(index)
+    }
+
+    /// Get the index of a validator in the committee.
+    pub fn index(&self, validator: &validator::PublicKey) -> Option<usize> {
+        self.indexes.get(validator).copied()
+    }
+
+    /// Computes the leader for the given view.
+    pub fn view_leader(
+        &self,
+        view_number: ViewNumber,
+        leader_selection: &LeaderSelectionMode,
+    ) -> validator::PublicKey {
+        match &leader_selection {
+            LeaderSelectionMode::RoundRobin => {
+                let index = view_number.0 as usize % self.len();
+                self.get(index).unwrap().key.clone()
+            }
+            LeaderSelectionMode::Weighted => {
+                let eligibility = LeaderSelectionMode::leader_weighted_eligibility(
+                    view_number.0,
+                    self.total_weight,
+                );
+                let mut offset = 0;
+                for val in &self.vec {
+                    offset += val.weight;
+                    if eligibility < offset {
+                        return val.key.clone();
+                    }
+                }
+                unreachable!()
+            }
+            LeaderSelectionMode::Sticky(pk) => {
+                let index = self.index(pk).unwrap();
+                self.get(index).unwrap().key.clone()
+            }
+            LeaderSelectionMode::Rota(pks) => {
+                let index = view_number.0 as usize % pks.len();
+                let index = self.index(&pks[index]).unwrap();
+                self.get(index).unwrap().key.clone()
+            }
+        }
+    }
+
+    /// Signature weight threshold for this validator committee.
+    pub fn quorum_threshold(&self) -> u64 {
+        quorum_threshold(self.total_weight())
+    }
+
+    /// Signature weight threshold for this validator committee to trigger a reproposal.
+    pub fn subquorum_threshold(&self) -> u64 {
+        subquorum_threshold(self.total_weight())
+    }
+
+    /// Maximal weight of faulty replicas allowed in this validator committee.
+    pub fn max_faulty_weight(&self) -> u64 {
+        max_faulty_weight(self.total_weight())
+    }
+
+    /// Compute the sum of signers weights.
+    /// Panics if signers length does not match the number of validators in committee
+    pub fn weight(&self, signers: &Signers) -> u64 {
+        assert_eq!(self.vec.len(), signers.len());
+        self.vec
+            .iter()
+            .enumerate()
+            .filter(|(i, _)| signers.0[*i])
+            .map(|(_, v)| v.weight)
+            .sum()
+    }
+
+    /// Sum of all validators' weight in the committee
+    pub fn total_weight(&self) -> u64 {
+        self.total_weight
+    }
+}
+
+/// Validator representation inside a Committee.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct WeightedValidator {
+    /// Validator key
+    pub key: validator::PublicKey,
```

Comment:
let's add proof-of-possession to this new version of the protocol.

Reply:
That would make it backwards incompatible with ENs, which I'm trying to avoid. Better to add it after we remove the validator committee from Genesis.